TypeError in warnings machinery

<!--

Thanks for opening an issue! To help the Numba team handle your information
efficiently, please first ensure that there is no other issue present that
already describes the issue you have
(search at https://github.com/numba/numba/issues?&q=is%3Aissue).

For more general "how do I do X?" type questions, please speak to us in real
time on https://gitter.im/numba/numba or post to the Numba mailing list
https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

-->

## Reporting a bug

<!--

Before submitting a bug report please ensure that you can check off these boxes:

-->

- [x] I am using the latest ~~released~~ master version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

<!--

Please include details of the bug here, including, if applicable, what you
expected to happen!

-->
I see the error below on master which I believe is due to the introduction of deprecated warnings. `WarningsFixer` tries to sort warnings [here](https://github.com/numba/numba/blob/master/numba/errors.py#L425), but Numba warnings don't have comparison implemented. This happens only if previous attributes of the warning data structure happen to be the same. I haven't been able to create a small reproducer for this yet.

```
../numba/numba/compiler.py:1017: in compile_ir
    lifted_from=lifted_from)
../numba/numba/compiler.py:385: in compile_ir
    return self._compile_ir()
../numba/numba/compiler.py:893: in _compile_ir
    return self._compile_core()
../numba/numba/compiler.py:873: in _compile_core
    res = pm.run(self.status)
../numba/numba/compiler_lock.py:32: in _acquire_compile_lock
    return func(*args, **kwargs)
../numba/numba/compiler.py:254: in run
    raise patched_exception
../numba/numba/compiler.py:245: in run
    stage()
../numba/numba/compiler.py:501: in stage_nopython_frontend
    self.locals)
../numba/numba/compiler.py:1109: in type_inference_stage
    warnings.flush()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <numba.errors.WarningsFixer object at 0x7f368b3c2cf8>

    def flush(self):
        """
        Emit all stored warnings.
        """
>       for (filename, lineno, category), messages in sorted(self._warnings.items()):
E       TypeError: Failed in bodo mode pipeline (step: convert to distributed)
E       Failed in nopython mode pipeline (step: nopython frontend)
E       '<' not supported between instances of 'type' and 'type'

../numba/numba/errors.py:425: TypeError
```

--- a/.flake8
+++ b/.flake8
@@ -221,7 +221,6 @@ exclude =
     numba/tests/test_return_values.py
     numba/tests/test_parallel_backend.py
     numba/tests/test_nrt.py
-    numba/tests/test_warnings.py
     numba/tests/test_mangling.py
     numba/tests/test_npdatetime.py
     numba/tests/test_fancy_indexing.py

--- a/numba/errors.py
+++ b/numba/errors.py
@@ -24,6 +24,7 @@ class NumbaWarning(Warning):
     """
     Base category for all Numba compiler warnings.
     """
+
     def __init__(self, msg, loc=None, highlighting=True, ):
         self.msg = msg
         self.loc = loc
@@ -422,7 +423,15 @@ class WarningsFixer(object):
         """
         Emit all stored warnings.
         """
-        for (filename, lineno, category), messages in sorted(self._warnings.items()):
+        def key(arg):
+            # It is possible through codegen to create entirely identical
+            # warnings, this leads to comparing types when sorting which breaks
+            # on Python 3. Key as str() and if the worse happens then `id`
+            # creates some uniqueness
+            return str(arg) + str(id(arg))
+
+        for (filename, lineno, category), messages in sorted(
+                self._warnings.items(), key=key):
             for msg in sorted(messages):
                 warnings.warn_explicit(msg, category, filename, lineno)
         self._warnings.clear()

--- a/numba/tests/test_warnings.py
+++ b/numba/tests/test_warnings.py
@@ -5,19 +5,21 @@ import numpy as np
 import numba.unittest_support as unittest
 from numba import jit
 from numba.errors import NumbaWarning, deprecated, NumbaDeprecationWarning
+from numba import errors
+
 
 class TestBuiltins(unittest.TestCase):
 
     def check_objmode_deprecation_warning(self, w):
         # Object mode fall-back is slated for deprecation, check the warning
         msg = ("Fall-back from the nopython compilation path to the object "
-                "mode compilation path has been detected")
+               "mode compilation path has been detected")
         self.assertEqual(w.category, NumbaDeprecationWarning)
         self.assertIn(msg, str(w.message))
 
     def test_type_infer_warning(self):
         def add(x, y):
-            a = {}
+            a = {} # noqa dead
             return x + y
 
         with warnings.catch_warnings(record=True) as w:
@@ -40,6 +42,7 @@ class TestBuiltins(unittest.TestCase):
 
     def test_return_type_warning(self):
         y = np.ones(4, dtype=np.float32)
+
         def return_external_array():
             return y
 
@@ -80,10 +83,9 @@ class TestBuiltins(unittest.TestCase):
             # No more warning
             self.assertEqual(len(w), 0)
 
-
     def test_no_warning_with_forceobj(self):
         def add(x, y):
-            a = []
+            a = [] # noqa dead
             return x + y
 
         with warnings.catch_warnings(record=True) as w:
@@ -96,7 +98,7 @@ class TestBuiltins(unittest.TestCase):
 
     def test_loop_lift_warn(self):
         def do_loop(x):
-            a = {}
+            a = {} # noqa dead
             for i in range(x.shape[0]):
                 x[i] *= 2
 
@@ -132,7 +134,8 @@ class TestBuiltins(unittest.TestCase):
 
     def test_deprecated(self):
         @deprecated('foo')
-        def bar(): pass
+        def bar():
+            pass
 
         with warnings.catch_warnings(record=True) as w:
             warnings.simplefilter('always')
@@ -142,6 +145,28 @@ class TestBuiltins(unittest.TestCase):
             self.assertEqual(w[0].category, DeprecationWarning)
             self.assertIn('bar', str(w[0].message))
             self.assertIn('foo', str(w[0].message))
+
+    def test_warnings_fixer(self):
+        # For some context, see #4083
+
+        wfix = errors.WarningsFixer(errors.NumbaWarning)
+        with wfix.catch_warnings('foo', 10):
+            warnings.warn(errors.NumbaWarning('same'))
+            warnings.warn(errors.NumbaDeprecationWarning('same'))
+
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter('always')
+            wfix.flush()
+
+            self.assertEqual(len(w), 2)
+            # the order of these will be backwards to the above, the
+            # WarningsFixer flush method sorts with a key based on str
+            # comparison
+            self.assertEqual(w[0].category, NumbaDeprecationWarning)
+            self.assertEqual(w[1].category, NumbaWarning)
+            self.assertIn('same', str(w[0].message))
+            self.assertIn('same', str(w[1].message))
+
 
 if __name__ == '__main__':
     unittest.main()

